File: src/bot/utils/splitMessage.ts
const PREFIX_RESERVE = 8 // "[99/99] "

/**
 * Splits text into chunks that fit within `limit` characters (default 4096).
 * Splits preferentially at \n\n paragraph boundaries, then at word boundaries,
 * then hard-cuts as a last resort.
 * When more than one chunk is produced, each chunk is prefixed [N/M].
 * Reserves 8 characters per chunk for the prefix to ensure prefixed chunks
 * never exceed `limit`.
 */
export function splitMessage(text: string, limit = 4096): string[] {
  if (text.length <= limit) return [text]

  const effectiveLimit = limit - PREFIX_RESERVE
  const chunks: string[] = []

  function pushLongText(str: string) {
    let remaining = str
    while (remaining.length > effectiveLimit) {
      const slice = remaining.slice(0, effectiveLimit)
      const lastSpace = slice.lastIndexOf(" ")
      if (lastSpace > 0) {
        chunks.push(remaining.slice(0, lastSpace))
        remaining = remaining.slice(lastSpace + 1)
      } else {
        // Hard cut — prevents infinite loop on text with no spaces
        chunks.push(remaining.slice(0, effectiveLimit))
        remaining = remaining.slice(effectiveLimit)
      }
    }
    return remaining
  }

  const paragraphs = text.split("\n\n")
  let current = ""

  for (const para of paragraphs) {
    if (para.length > effectiveLimit) {
      if (current) {
        chunks.push(current)
        current = ""
      }
      current = pushLongText(para)
    } else {
      const candidate = current ? `${current}\n\n${para}` : para
      if (candidate.length > effectiveLimit) {
        if (current) chunks.push(current)
        current = para
      } else {
        current = candidate
      }
    }
  }

  if (current) chunks.push(current)

  if (chunks.length > 1) {
    const total = chunks.length
    return chunks.map((chunk, i) => `[${i + 1}/${total}] ${chunk}`)
  }

  return chunks
}

File: src/bot/utils/split_message.rs
/// Reserve space for chunk prefix like `[1/9] `.
const PREFIX_RESERVE: usize = 8;

/// Split a potentially long text into Telegram-safe chunks (≤ `limit` chars each).
///
/// - Tries to split on paragraph boundaries (`\n\n`) first.
/// - Falls back to splitting on spaces, then hard-splits if necessary.
/// - When more than one chunk is produced, every chunk is prefixed `[N/Total] `.
pub fn split_message(text: &str, limit: usize) -> Vec<String> {
    if text.len() <= limit {
        return vec![text.to_string()];
    }

    let effective = limit.saturating_sub(PREFIX_RESERVE);

    let mut chunks: Vec<String> = Vec::new();

    /// Push a string that may be longer than `effective`, splitting on spaces
    /// or hard-splitting as a last resort.  Returns whatever didn't fit.
    fn push_long_text(s: &str, effective: usize, chunks: &mut Vec<String>) -> String {
        let mut remaining = s.to_string();
        while remaining.len() > effective {
            let slice = &remaining[..effective];
            let split_at = slice.rfind(' ').filter(|&p| p > 0).unwrap_or(effective);
            chunks.push(remaining[..split_at].to_string());
            remaining = remaining[split_at..].trim_start().to_string();
        }
        remaining
    }

    let paragraphs: Vec<&str> = text.split("\n\n").collect();
    let mut current = String::new();

    for para in &paragraphs {
        if para.len() > effective {
            // Flush current buffer first
            if !current.is_empty() {
                chunks.push(std::mem::take(&mut current));
            }
            current = push_long_text(para, effective, &mut chunks);
        } else {
            let candidate = if current.is_empty() {
                para.to_string()
            } else {
                format!("{}\n\n{}", current, para)
            };

            if candidate.len() > effective {
                if !current.is_empty() {
                    chunks.push(std::mem::take(&mut current));
                }
                current = para.to_string();
            } else {
                current = candidate;
            }
        }
    }

    if !current.is_empty() {
        chunks.push(current);
    }

    if chunks.len() <= 1 {
        return chunks;
    }

    let total = chunks.len();
    chunks
        .into_iter()
        .enumerate()
        .map(|(i, chunk)| format!("[{}/{}] {}", i + 1, total, chunk))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn short_message_is_unchanged() {
        let msg = "Hello world";
        assert_eq!(split_message(msg, 4096), vec![msg.to_string()]);
    }

    #[test]
    fn long_message_gets_numbered() {
        let para = "word ".repeat(200); // ~1000 chars
        let text = format!("{}\n\n{}", para.trim(), para.trim());
        let parts = split_message(&text, 300);
        assert!(parts.len() > 1);
        assert!(parts[0].starts_with("[1/"));
    }
}
